File: 0416/recursion.rs
struct Solution;

impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let sum: i32 = nums.iter().sum();
        if sum % 2 == 1 { return false; }
        let target: i32 = sum / 2;
        let target_usize : usize = target as usize;
        let mut dp: Vec<Vec<i32>> = vec![vec![-1; target_usize+1]; nums.len()];
        return Solution::recursion(&nums, 0, target, &mut dp);
    }

    pub fn recursion(nums: &Vec<i32>, idx: usize, target: i32, dp: &mut Vec<Vec<i32>>) -> bool {
        if target == 0 {
            return true;
        } else if target < 0 {
            return false;
        } else if idx == nums.len() {
            return false;
        }
        let target_usize : usize = target as usize;
        if dp[idx][target_usize] == 0 { return true; }
        else if dp[idx][target_usize] == 1 { return false; }
        let ret: bool = Solution::recursion(&nums, idx+1, target, dp) ||
            Solution::recursion(&nums, idx+1, target - nums[idx], dp);
        if ret { dp[idx][target_usize] = 0; }
        else { dp[idx][target_usize] = 1; }
        return ret;
    }
}

fn main() {
    let mut nums: Vec<i32>;

    nums = vec![1,5,11,5];
    assert!(Solution::can_partition(nums));

    nums = vec![1,2,3,5];
    assert!(!Solution::can_partition(nums));
}

File: 0416/recursion.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2 == 1:
            return False
        target = int(total / 2)
        # idx x target
        # -1: unvisited, 0: False, 1: True
        memo = [[-1 for _ in range(target)] for idx in range(len(nums))]
        return self.recursion(nums, 0, 0, target, memo)

    def recursion(self, nums, idx, current_total, target, memo):
        if current_total == target:
            return True
        elif current_total > target:
            return False
        if idx >= len(nums):
            return False
        if memo[idx][current_total] != -1:
            return True if memo[idx][current_total] == 1 else False
        can_partition = self.recursion(nums, idx+1, current_total + nums[idx], target, memo) or self.recursion(nums, idx+1, current_total, target, memo)
        memo[idx][current_total] = 1 if can_partition else 0
        return memo[idx][current_total]


if __name__ == '__main__':
    sol = Solution()

    nums = [1, 5, 11, 5]
    assert sol.canPartition(nums)

    nums = [1, 2, 3, 5]
    assert not sol.canPartition(nums)
